### PR TITLE
feat(save-user-progress)

### PR DESCRIPTION
**Pull request by emberglaz**
* rename quest.ts -> initialQuest.ts
* add questSaga
* add plantSaga
* modify plantSlice + actions
* modify questSlice + actions
* add firestoreCreateUserProgress
* add firestoreUpdareUserProgress

### CODE REVIEW

**Comment by emberglaz** [comment]
You can only update completely all states

File: app/constants/progress.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+export const INITIAL_PLANTS_STACK = [
+  PlantsType.SpourCompassion,
+  PlantsType.SpourCompassion,
+  PlantsType.SpourCompassion,
+  PlantsType.SpourCompassion,
+  PlantsType.SpourCompassion,
+];
+export const INITIAL_PLANT_AREA = {
+  TopLeft: null,
```
Comment (Reykjavik151) [inline app/constants/progress.ts]: spaceline

File: app/redux/slices/questSlice.ts [hunk 2]
```diff
@@ -52,6 +52,25 @@ export const questSlice = createSlice({
...
+          interruptedQuestLine,
+        },
+      }: PayloadAction<IQuestProgress>,
+    ) {
+      state.completedQuestsId = completedQuestsId;
+      state.currentDay = currentDay;
+      state.currentDayQuestsStack = currentDayQuestsStack;
+      state.lastDayUpdate = lastDayUpdate;
+      state.interruptedQuestLine = interruptedQuestLine;
+    },
```
Comment (Reykjavik151) [inline app/redux/slices/questSlice.ts]:
```
setQuestState(
state,
{ payload },
) {
_.merge(state, payload);
}

File: app/sagas/quest.ts [hunk 3]
```diff
@@ -1,30 +1,193 @@
...
+  );
+  const currentDayQuestsStack: number[] = yield select(
+    state => state.quest.currentDayQuestsStack,
+  );
+  const lastDayUpdate: number = yield select(
+    state => state.quest.lastDayUpdate,
+  );
+  const interruptedQuestLine: Nullable<IInterruptedQuestLine> = yield select(
+    state => state.quest.interruptedQuestLine,
+  );
```
Comment (Reykjavik151) [inline app/sagas/quest.ts]: create `selector` function, and use it as
```
const blablabla: SelectorReturnObjectName = yield select(selectorFunction);
```

File: app/sagas/quest.ts [hunk 4]
```diff
@@ -1,30 +1,193 @@
...
+  const completedQuestsId: number[] = yield select(
+    state => state.quest.completedQuestsId,
```
Comment (Reykjavik151) [inline app/sagas/quest.ts]: the same as above

File: app/sagas/quest.ts [hunk 5]
```diff
@@ -1,30 +1,193 @@
...
+  );
+  const currentDay: number = yield select(state => state.quest.currentDay);
+  const currentDayQuestsStack: number[] = yield select(
+    state => state.quest.currentDayQuestsStack,
+  );
+  const lastDayUpdate: number = yield select(
+    state => state.quest.lastDayUpdate,
+  );
+
```
Comment (Reykjavik151) [inline app/sagas/quest.ts]: the same as above

File: app/sagas/quest.ts [hunk 6]
```diff
@@ -1,30 +1,193 @@
...
+  const currentDay: number = yield select(state => state.quest.currentDay);
+  const currentDayQuestsStack: number[] = yield select(
+    state => state.quest.currentDayQuestsStack,
+  );
+  const lastDayUpdate: number = yield select(
+    state => state.quest.lastDayUpdate,
+  );
+  const interruptedQuestLine: Nullable<IInterruptedQuestLine> = yield select(
+    state => state.quest.interruptedQuestLine,
+  );
```
Comment (Reykjavik151) [inline app/sagas/quest.ts]: the same as above

File: app/sagas/quest.ts [hunk 7]
```diff
@@ -1,30 +1,193 @@
...
+  const completedQuestsId: number[] = yield select(
+    state => state.quest.completedQuestsId,
+  );
+  const currentDay: number = yield select(state => state.quest.currentDay);
+  const currentDayQuestsStack: number[] = yield select(
+    state => state.quest.currentDayQuestsStack,
+  );
+  const interruptedQuestLine: Nullable<IInterruptedQuestLine> = yield select(
+    state => state.quest.interruptedQuestLine,
+  );
```
Comment (Reykjavik151) [inline app/sagas/quest.ts]: the same as above

File: app/screens/MixingElixir/MixingElixirSuccess/MixingElixirSuccess.tsx [hunk 8]
```diff
@@ -20,7 +20,10 @@ export const MixingElixirSuccessScreen: React.FC<IMixingElixirSuccessScreenProps
     const dispatch = useAppDispatch();
 
     const currentPlant = useAppSelector(
-      state => state.plant.plantsStack[state.plant.plantsStack.length - 1],
+      state =>
+        state.plant.plantsStack?.[
+          state.plant.plantsStack.length - 1
+        ] as PlantsType,
```
Comment (Reykjavik151) [inline app/screens/MixingElixir/MixingElixirSuccess/MixingElixirSuccess.tsx]: if plantsStack is empty --> error
refactor

**Review by Reykjavik151 — CHANGES_REQUESTED** [review]

File: app/sagas/quest.ts [hunk 9]
```diff
@@ -1,30 +1,152 @@
...
+};
 
-  for (const language of locales) {
-    const quests: IQuestDatoCms[] = yield call(
-      datocmsApiInstance.getAllQuests,
-      language,
+function* watchUpdateCurrentDay({ payload }: PayloadAction<number>) {
+  const currentQuestState: IQuestProgress = yield select(state =>
+    questStateSelector(state),
+  );
```
Comment (Reykjavik151) [inline app/sagas/quest.ts]: `yield select(questStateSelector);`

**Review by Reykjavik151 — CHANGES_REQUESTED** [review]

**Pull request merged**